Notify networking stack that download is media related

jyavenard opened the pull request:
#### 127556767fd6049032090249e4ccc91110e3c70f
<pre>
Notify networking stack that download is media related
 <a href="https://bugs.webkit.org/show_bug.cgi?id=249569">https://bugs.webkit.org/show_bug.cgi?id=249569</a> rdar://103046339

Reviewed by Youenn Fablet.

Send a system notification when a networking operation is likely
going to be related to media playback.
For now the heuristic is rather primitive: if we have an active source buffer
linked to a media element we assume that all further networking operation in the
page is media related.
The notification is produced by the network process whenever a content process
has an active source buffers, and when all have been deleted.

The code was manually tested on a physical iPhone, however it fails on the simulator.
The api test will be added in a follow-up change (<a href="https://bugs.webkit.org/show_bug.cgi?id=250420)">https://bugs.webkit.org/show_bug.cgi?id=250420)</a>
* Source/WebCore/Modules/mediasource/MediaSource.cpp:
(WebCore::MediaSource::removeSourceBuffer):
(WebCore::MediaSource::regenerateActiveSourceBuffers):
(WebCore::MediaSource::notifyElementUpdateMediaState const):
* Source/WebCore/Modules/mediasource/MediaSource.h:
* Source/WebCore/html/HTMLMediaElement.cpp:
(WebCore::HTMLMediaElement::updateMediaState):
(WebCore::HTMLMediaElement::mediaState const):
* Source/WebCore/html/HTMLMediaElement.h:
* Source/WebCore/page/MediaProducer.h:
* Source/WebKit/NetworkProcess/NetworkProcess.cpp:
(WebKit::NetworkProcess::didClose):
(WebKit::NetworkProcess::notifyMediaStreamingActivity):
* Source/WebKit/NetworkProcess/NetworkProcess.h:
* Source/WebKit/NetworkProcess/NetworkProcess.messages.in:
* Source/WebKit/UIProcess/Cocoa/WebProcessPoolCocoa.mm:
* Source/WebKit/UIProcess/Network/NetworkProcessProxy.cpp:
(WebKit::NetworkProcessProxy::notifyMediaStreamingActivity):
* Source/WebKit/UIProcess/Network/NetworkProcessProxy.h:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::updatePlayingMediaDidChange):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/WebProcessPool.cpp:
(WebKit::m_webProcessWithMediaStreamingCounter):
(WebKit::WebProcessPool::webProcessWithMediaStreamingToken const):
(WebKit::WebProcessPool::updateMediaStreamingActivity):
(WebKit::WebProcessPool::notifyMediaStreamingActivity):
* Source/WebKit/UIProcess/WebProcessPool.h:
* Source/WebKit/UIProcess/WebProcessProxy.cpp:
(WebKit::WebProcessProxy::shutDown):
(WebKit::WebProcessProxy::removeWebPage):
(WebKit::WebProcessProxy::updateAudibleMediaAssertions):
(WebKit::WebProcessProxy::updateMediaStreamingActivity):
* Source/WebKit/UIProcess/WebProcessProxy.h:

Canonical link: <a href="https://commits.webkit.org/258765@main">https://commits.webkit.org/258765@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/6e2033d6c3e0dcdfdbc2eabb53cd0a2652702218

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/102872 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/11990 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/35898 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/112125 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/172347 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/106835 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/13009 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/2898 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/95121 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/109792 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/108646 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/9991 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/93187 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/37632 "Passed tests") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/91836 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/24719 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/79367 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/5443 "Built successfully") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/70/builds/26133 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/5590 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/84/builds/2586 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/11607 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/45625 "Passed tests") | | 
| [❌ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/6023 "Failed to push commit to Webkit repository") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/7337 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

youennf commented on Source/WebKit/NetworkProcess/NetworkProcess.cpp:
```diff
@@ -1366,6 +1371,32 @@ bool NetworkProcess::privateClickMeasurementEnabled() const
+    if (m_mediaStreamingActivitityToken == NOTIFY_TOKEN_INVALID) {
+        status = notify_register_check(notifyMediaStreamingName, &m_mediaStreamingActivitityToken);
+        if (status != NOTIFY_STATUS_OK || m_mediaStreamingActivitityToken == NOTIFY_TOKEN_INVALID)
+            RELEASE_LOG_ERROR(IPC, "notify_register_check() for %s failed with status (%d) 0x%X", notifyMediaStreamingName, status, status);
```
RELEASE_LOG_ERROR_IF

jyavenard replied:
I'll take the route of having the UIProcess resend the notification if the network process crashes, I will take this in a follow-up bug.

youennf commented on Source/WebKit/NetworkProcess/NetworkProcess.cpp:
```diff
@@ -1366,6 +1371,32 @@ bool NetworkProcess::privateClickMeasurementEnabled() const
+        else {
+            status = notify_post(notifyMediaStreamingName);
+            if (status != NOTIFY_STATUS_OK)
+                RELEASE_LOG_ERROR(IPC, "notify_post() for %s failed with status (%d) 0x%X", notifyMediaStreamingName, status, status);
```
Ditto

youennf commented on Source/WebKit/NetworkProcess/NetworkProcess.cpp:
```diff
@@ -1366,6 +1371,32 @@ bool NetworkProcess::privateClickMeasurementEnabled() const
+
+    uint32_t status = NOTIFY_STATUS_FAILED;
+    if (m_mediaStreamingActivitityToken == NOTIFY_TOKEN_INVALID) {
+        status = notify_register_check(notifyMediaStreamingName, &m_mediaStreamingActivitityToken);
```
auto status

youennf commented on Source/WebKit/NetworkProcess/NetworkProcess.cpp:
```diff
@@ -1366,6 +1371,32 @@ bool NetworkProcess::privateClickMeasurementEnabled() const
+            RELEASE_LOG_ERROR(IPC, "notify_register_check() for %s failed with status (%d) 0x%X", notifyMediaStreamingName, status, status);
+    }
+    if (m_mediaStreamingActivitityToken != NOTIFY_TOKEN_INVALID) {
+        status = notify_set_state(m_mediaStreamingActivitityToken, activity ? 1 : 0);
```
auto status

youennf commented on Source/WebKit/NetworkProcess/NetworkProcess.cpp:
```diff
@@ -1366,6 +1371,32 @@ bool NetworkProcess::privateClickMeasurementEnabled() const
+        status = notify_register_check(notifyMediaStreamingName, &m_mediaStreamingActivitityToken);
+        if (status != NOTIFY_STATUS_OK || m_mediaStreamingActivitityToken == NOTIFY_TOKEN_INVALID)
+            RELEASE_LOG_ERROR(IPC, "notify_register_check() for %s failed with status (%d) 0x%X", notifyMediaStreamingName, status, status);
+    }
```
We could early return here in case of error so that we do not need the next if check.

youennf reviewed (approved):
r=me.

In case of network process crash, it is unclear whether the notification value will be correct or not (the IPC message might be lost due to the crash).
We could either ensure this value is ok with an async reply that returns a success value.
Or maybe resend the same message if the activity was ever set previously.
This would require a state in NetworkProcessProxy.

jyavenard committed:
Notify networking stack that download is media related
 https://bugs.webkit.org/show_bug.cgi?id=249569 rdar://103046339

Reviewed by Youenn Fablet.

Send a system notification when a networking operation is likely
going to be related to media playback.
For now the heuristic is rather primitive: if we have an active source buffer
linked to a media element we assume that all further networking operation in the
page is media related.
The notification is produced by the network process whenever a content process
has an active source buffers, and when all have been deleted.

The code was manually tested on a physical iPhone, however it fails on the simulator.
The api test will be added in a follow-up change (https://bugs.webkit.org/show_bug.cgi?id=250420)
* Source/WebCore/Modules/mediasource/MediaSource.cpp:
(WebCore::MediaSource::removeSourceBuffer):
(WebCore::MediaSource::regenerateActiveSourceBuffers):
(WebCore::MediaSource::notifyElementUpdateMediaState const):
* Source/WebCore/Modules/mediasource/MediaSource.h:
* Source/WebCore/html/HTMLMediaElement.cpp:
(WebCore::HTMLMediaElement::updateMediaState):
(WebCore::HTMLMediaElement::mediaState const):
* Source/WebCore/html/HTMLMediaElement.h:
* Source/WebCore/page/MediaProducer.h:
* Source/WebKit/NetworkProcess/NetworkProcess.cpp:
(WebKit::NetworkProcess::didClose):
(WebKit::NetworkProcess::notifyMediaStreamingActivity):
* Source/WebKit/NetworkProcess/NetworkProcess.h:
* Source/WebKit/NetworkProcess/NetworkProcess.messages.in:
* Source/WebKit/UIProcess/Cocoa/WebProcessPoolCocoa.mm:
* Source/WebKit/UIProcess/Network/NetworkProcessProxy.cpp:
(WebKit::NetworkProcessProxy::notifyMediaStreamingActivity):
* Source/WebKit/UIProcess/Network/NetworkProcessProxy.h:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::updatePlayingMediaDidChange):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/WebProcessPool.cpp:
(WebKit::m_webProcessWithMediaStreamingCounter):
(WebKit::WebProcessPool::webProcessWithMediaStreamingToken const):
(WebKit::WebProcessPool::updateMediaStreamingActivity):
(WebKit::WebProcessPool::notifyMediaStreamingActivity):
* Source/WebKit/UIProcess/WebProcessPool.h:
* Source/WebKit/UIProcess/WebProcessProxy.cpp:
(WebKit::WebProcessProxy::shutDown):
(WebKit::WebProcessProxy::removeWebPage):
(WebKit::WebProcessProxy::updateAudibleMediaAssertions):
(WebKit::WebProcessProxy::updateMediaStreamingActivity):
* Source/WebKit/UIProcess/WebProcessProxy.h:

Canonical link: https://commits.webkit.org/258765@main
